server/src/main/java/com/exacaster/lighter/backend/yarn/YarnBackend.java

```diff
@@ -53,26 +46,37 @@ public YarnBackend(YarnProperties yarnProperties, YarnClient client, AppConfigur
     }
 
     private ApplicationState getState(String id) {
-        var yarnApplication = client.getApplication(id).getApp();
-        switch (yarnApplication.getFinalStatus()) {
-            case "UNDEFINED":
-                return ApplicationState.BUSY;
-            case "SUCCEEDED":
-                return ApplicationState.SUCCESS;
-            case "FAILED":
-                return ApplicationState.ERROR;
-            case "KILLED":
-                return ApplicationState.KILLED;
-            default:
-                throw new IllegalStateException("Unexpected state: " + yarnApplication);
+        try {
+            var yarnApplication = client.getApplicationReport(fromString(id));
+            switch (yarnApplication.getFinalApplicationStatus()) {
+                case UNDEFINED:
+                    return ApplicationState.BUSY;
+                case SUCCEEDED:
+                    return ApplicationState.SUCCESS;
+                case FAILED:
+                    return ApplicationState.ERROR;
+                case KILLED:
+                    return ApplicationState.KILLED;
+            }
+        } catch (YarnException | IOException e) {
+            LOG.error("Unexpected error for appId: {}", id, e);
         }
+        throw new IllegalStateException("Unexpected state for appId: " + id);
     }
 
     @Override
     public Optional<Log> getLogs(Application application) {
         // TODO: extract yarn logs, returning tracking url for now
-        return getYarnApplicationId(application).map(client::getApplication)
-                .map(YarnApplicationResponse::getApp)
+        return getYarnApplicationId(application)
+                .map(ApplicationId::fromString)
+                .map(appId -> {
+                    try {
+                        return client.getApplicationReport(appId);
+                    } catch (YarnException | IOException e) {
+                        LOG.error("Failed to get logs for app: {}", application, e);
+                        return null;
+                    }
+                })
                 .map(a -> new Log(application.getId(), a.getTrackingUrl()));
```

Comment:
this will cause NullPointerException, since you sometimes return null in previous `map` call.

Reply:
optional will be empty - added test for it